sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs

```diff
@@ -1115,23 +1101,64 @@ private void StopTrackingLinkAsActive(AmqpObject link, Timer authorizationRefres
         }
 
         /// <summary>
-        ///   Performs the actions needed to open an AMQP object, such
-        ///   as a session or link for use.
+        ///   Performs the actions needed to open an AMQP link.
         /// </summary>
-        ///
-        /// <param name="target">The target AMQP object to open.</param>
-        /// <param name="timeout">The timeout to apply when opening the link.</param>
+        /// <param name="link">The target AMQP object to open.</param>
+        /// <param name="entityPath">The path of the entity associated with the AMQP object being opened, if any.</param>
         /// <param name="cancellationToken">Token to signal cancellation of the operation.</param>
+        protected virtual async Task OpenAmqpLinkAsync(
+            AmqpLink link,
+            string entityPath,
+            CancellationToken cancellationToken)
+        {
+            await OpenAmqpObjectCoreAsync(link, entityPath, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        ///  Performs the actions needed to open a request response link. This overload is necessary because <see cref="RequestResponseAmqpLink"/>
+        ///  does not inherit from <see cref="AmqpLink"/>.
+        /// </summary>
+        /// <param name="link">The target AMQP object to open.</param>
         /// <param name="entityPath">The path of the entity associated with the AMQP object being opened, if any.</param>
+        /// <param name="cancellationToken">Token to signal cancellation of the operation.</param>
+        protected virtual async Task OpenAmqpLinkAsync(
+            RequestResponseAmqpLink link,
+            string entityPath,
+            CancellationToken cancellationToken)
+        {
+            await OpenAmqpObjectCoreAsync(link, entityPath, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        ///  Performs the actions needed to open an AmqpObject.
+        /// </summary>
+        /// <param name="targetObject">The target AMQP object to open.</param>
+        /// <param name="timeout">The timeout to apply when opening the object.</param>
         protected virtual async Task OpenAmqpObjectAsync(
+            AmqpObject targetObject,
+            TimeSpan timeout)
+        {
+            await OpenAmqpObjectCoreAsync(targetObject, timeout: timeout).ConfigureAwait(false);
+        }
+
+        private static async Task OpenAmqpObjectCoreAsync(
             AmqpObject target,
-            TimeSpan timeout,
-            CancellationToken cancellationToken,
-            string entityPath = default)
+            string entityPath = default,
+            TimeSpan? timeout = default,
+            CancellationToken? cancellationToken = default)
         {
+            // only one of timeout or cancellation token should be set
+            Debug.Assert(timeout.HasValue ^ cancellationToken.HasValue);
```

Comment:
Just curious, why Debug.Assert instead of Assert?

Reply:
Why an assert here at all?

Reply:
This is meant to check for coding errors so by using Debug.Assert it wouldn't actually run in a release build but we would still be able to catch any mistakes in our debug builds.

Reply:
> Why an assert here at all?

I suppose it is probably not needed now that I refactored this to be a private method. Initially I had all parameters in the public method. Still it probably doesn't hurt in documenting the intent, and in case we add a new method that has both params.